zcash_client_sqlite/src/wallet.rs

```diff
@@ -732,166 +760,322 @@ pub(crate) fn get_transparent_balances<P: consensus::Parameters>(
+        ":to_address": &to_address,
+        ":to_account": &to_account,
+        ":value": &i64::from(output.value()),
+        ":memo": output.memo().filter(|m| *m != &MemoBytes::empty()).map(|m| m.as_slice()),
+    ];
+
+    stmt_insert_sent_output.execute(sql_args)?;
+
+    Ok(())
 }
 
-/// Records information about a transaction output that your wallet created.
+/// Records information about a transaction output that your wallet created, from the constituent
+/// properties of that output.
 ///
-/// This is a crate-internal convenience method.
+/// - If `recipient` is a Unified address, `output_index` is an index into the outputs of the
+///   transaction within the bundle associated with the recipient's output pool.
+/// - If `recipient` is a Sapling address, `output_index` is an index into the Sapling outputs of
+///   the transaction.
+/// - If `recipient` is a transparent address, `output_index` is an index into the transparent
+///   outputs of the transaction.
+/// - If `recipient` is an internal account, `output_index` is an index into the Sapling outputs of
+///   the transaction.
 #[allow(clippy::too_many_arguments)]
-pub(crate) fn put_sent_output<'a, P: consensus::Parameters>(
-    stmts: &mut DataConnStmtCache<'a, P>,
+pub(crate) fn put_sent_output<P: consensus::Parameters>(
+    conn: &rusqlite::Connection,
+    params: &P,
     from_account: AccountId,
     tx_ref: i64,
     output_index: usize,
     recipient: &Recipient,
     value: Amount,
     memo: Option<&MemoBytes>,
 ) -> Result<(), SqliteClientError> {
-    if !stmts.stmt_update_sent_output(from_account, recipient, value, memo, tx_ref, output_index)? {
-        stmts.stmt_insert_sent_output(
-            tx_ref,
-            output_index,
-            from_account,
-            recipient,
-            value,
-            memo,
-        )?;
-    }
+    let mut stmt_upsert_sent_output = conn.prepare_cached(
+        "INSERT INTO sent_notes (
+            tx, output_pool, output_index, from_account,
+            to_address, to_account, value, memo)
+        VALUES (
+            :tx, :output_pool, :output_index, :from_account,
+            :to_address, :to_account, :value, :memo)
+        ON CONFLICT (tx, output_pool, output_index) DO UPDATE
+        SET from_account = :from_account,
+            to_address = :to_address,
+            to_account = :to_account,
+            value = :value,
+            memo = IFNULL(:memo, memo)",
+    )?;
+
+    let (to_address, to_account, pool_type) = recipient_params(params, recipient);
+    let sql_args = named_params![
+        ":tx": &tx_ref,
+        ":output_pool": &pool_code(pool_type),
+        ":output_index": &i64::try_from(output_index).unwrap(),
+        ":from_account": &u32::from(from_account),
+        ":to_address": &to_address,
+        ":to_account": &to_account,
+        ":value": &i64::from(value),
+        ":memo": &memo.filter(|m| *m != &MemoBytes::empty()).map(|m| m.as_slice()),
```

Comment:
`memo.filter(|m| *m != &MemoBytes::empty()).map(|m| m.as_slice())` is used three times I think; consider extracting it to a function.